calyx/src/passes/group_to_invoke.rs

```diff
@@ -31,100 +34,190 @@ impl Named for GroupToInvoke {
     }
 }
 
+// Returns true if port's parent is cell. We can do this by checking if they
+// have the same name.
+fn cell_is_parent(port: &ir::Port, cell: &ir::RRC<ir::Cell>) -> bool {
+    port.get_parent_name() == cell.borrow().name()
+}
+
 /// Construct an [ir::Invoke] from an [ir::Group] that has been validated by this pass.
 fn construct_invoke(
     assigns: &[ir::Assignment],
     comp: RRC<ir::Cell>,
+    builder: &mut ir::Builder,
 ) -> ir::Control {
     let mut inputs = Vec::new();
-    let mut outputs = Vec::new();
+    let mut comb_assigns = Vec::new();
 
-    let cell_is_parent = |port: &ir::Port| -> bool {
-        if let ir::PortParent::Cell(cell_wref) = &port.parent {
-            Rc::ptr_eq(&cell_wref.upgrade(), &comp)
-        } else {
-            false
+    // Check if port's parent is a combinational primitive
+    let comb_is_parent = |port: &ir::Port| -> bool {
+        if !port.is_hole() {
+            if let ir::CellType::Primitive { is_comb, .. } =
+                port.cell_parent().borrow().prototype
+            {
+                return is_comb;
+            }
         }
+        false
     };
 
     for assign in assigns {
-        // If the cell's port is being used as a source, add the dst to
-        // outputs
-        if cell_is_parent(&assign.src.borrow())
-            && assign.src != comp.borrow().get_with_attr("done")
-        {
-            let name = assign.src.borrow().name.clone();
-            outputs.push((name, Rc::clone(&assign.dst)));
+        // We know that all assignments in this group should write to either a)
+        // a combinational component or b) comp or c) the group's done port
```

Comment:
Is this a precondition on the `assigns` array?

Reply:
Yes, since before each call of this method, we check for this (among other things) and if it is not satisfied then we don't call the method and instead just `continue`